src/main/core/logger/shadow_logger.rs

```diff
@@ -0,0 +1,426 @@
+                line = record
+                    .line
+                    .map(|l| format!("{}", l))
+                    .as_ref()
+                    .map(|s| s.as_str())
+                    .unwrap_or("n/a"),
+                module = record.module_path.unwrap_or("n/a"),
+                msg = record.message
+            )?;
+        }
+        if let Some(done_sender) = done_sender {
+            // We can't log from this thread without risking deadlock, so in the
+            // unlikely case that the calling thread has gone away, just print
+            // directly.
+            done_sender.send(()).unwrap_or_else(|e| {
+                println!(
+                    "WARNING: Logger couldn't notify
+                calling thread: {:?}",
+                    e
+                )
+            });
+        }
+        Ok(())
+    }
+
+    /// When disabled, the logger thread is notified to write each record as
+    /// soon as it's created.  The calling thread still isn't blocked on the
+    /// record actually being written, though.
+    pub fn set_buffering_enabled(&self, buffering_enabled: bool) {
+        let mut writer = self.buffering_enabled.write().unwrap();
+        *writer = buffering_enabled;
+    }
+
+    // Send a flush command to the logger thread.
+    fn flush_impl(&self, notify_done: Option<Sender<()>>) {
+        self.send_command(LoggerCommand::Flush(notify_done))
+    }
+
+    // Send a flush command to the logger thread and block until it's completed.
+    fn flush_sync(&self) {
+        let (done_sender, done_receiver) = std::sync::mpsc::channel();
+        self.flush_impl(Some(done_sender));
+        done_receiver.recv().unwrap();
+    }
+
+    // Send a flush command to the logger thread.
+    fn flush_async(&self) {
+        self.flush_impl(None);
+    }
+
+    // Send a command to the logger thread.
+    fn send_command(&self, cmd: LoggerCommand) {
+        SENDER
+            .try_with(|thread_sender| {
+                if thread_sender.borrow().is_none() {
+                    let lock = self.command_sender.lock().unwrap();
+                    *thread_sender.borrow_mut() = Some(lock.clone());
+                }
+                thread_sender
+                    .borrow()
+                    .as_ref()
+                    .unwrap()
+                    .send(cmd)
+                    .unwrap_or_else(|e| {
+                        println!("WARNING: Couldn't send command to logger thread: {:?}", e);
+                    });
+            })
+            .unwrap_or_else(|e| {
+                println!(
+                    "WARNING: Couldn't get sender channel to logger thread: {:?}",
+                    e
+                );
+            });
+    }
+}
+
+impl Log for ShadowLogger {
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        metadata.level() <= log::max_level()
+    }
+
+    fn log(&self, record: &Record) {
+        if !self.enabled(record.metadata()) {
+            return;
+        }
```

Comment:
Is the individual host log level taken into account anywhere?

Reply:
Good catch! It is now :)